cpp/src/sampling/detail/sampling_utils_impl.cuh

```diff
@@ -0,0 +1,886 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cugraph/detail/decompress_edge_partition.cuh>
+#include <cugraph/edge_partition_device_view.cuh>
+#include <cugraph/partition_manager.hpp>
+#include <cugraph/utilities/device_comm.cuh>
+#include <cugraph/utilities/host_scalar_comm.cuh>
+
+#include <raft/handle.hpp>
+
+#include <thrust/binary_search.h>
+#include <thrust/distance.h>
+#include <thrust/for_each.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/remove.h>
+#include <thrust/sort.h>
+#include <thrust/tabulate.h>
+
+#include <rmm/device_uvector.hpp>
+
+#include <numeric>
+#include <vector>
+
+namespace cugraph {
+namespace detail {
+
+template <typename GraphViewType>
+rmm::device_uvector<typename GraphViewType::edge_type> compute_local_major_degrees(
+  raft::handle_t const& handle, GraphViewType const& graph_view)
+{
```

Comment:
So, this code is pretty much https://github.com/rapidsai/cugraph/blob/branch-22.06/cpp/src/structure/graph_view_impl.cuh#L88 less `col_comm.reduce(...)` (https://github.com/rapidsai/cugraph/blob/branch-22.06/cpp/src/structure/graph_view_impl.cuh#L155).

Better to re-factor (or at least FIXME) to avoid code-duplication?

Reply:
I'll add a FIXME.

I actually think much of this logic should be moved into the graph_view, we assume too much regarding implementation in these functions.